### ISSUE TITLE
Vercel only: 0.33.0: error Could not load the "sharp" module using the linux-x64 runtime

### ISSUE THREAD

**Issue by JClackett** [lovell/sharp#3870]
### Is this a possible bug in a feature of sharp, unrelated to installation?

- [x] Running `npm install sharp` completes without error.
- [x] Running `node -e "require('sharp')"` completes without error.

If you cannot confirm both of these, please open an [installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) instead.

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

If you cannot confirm this, please upgrade to the latest version and try again before opening an issue.

If you are using another package which depends on a version of `sharp` that is not the latest, please open an issue against that package instead.

### What are the steps to reproduce?

Running sharp on Vercel breaks after updating from 0.32.6 to 0.33.0

```
ERROR	Error: Could not load the "sharp" module using the linux-x64 runtime
```

Downgrading to 0.32.6 for now

**Comment by lovell** [comment]
Do you only see this problem on Vercel? If so, please can you contact their support team. If not, please can you provide answers to the questions in the [installation issue template](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md).

**Comment by c100k** [comment]
👋🏽 Not only on Vercel since I've just had the issue locally on my Mac, after upgrading.

```sh
Error: Could not load the "sharp" module using the darwin-x64 runtime
Possible solutions:
- Add platform-specific dependencies:
    npm install --os=darwin --cpu=x64 sharp
  or
    npm install --force @img/sharp-darwin-x64
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
```

**Comment by lovell** [comment]
@c100k Please open [a new installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) and answer all of the questions.

**Comment by SeongwoonHong** [comment]
any updates? I'm getting the same error

**Comment by JClackett** [comment]
I'm pretty sure it's just that the new version has stricter OS/cpu requirements and likely that the installs that's we're used to using aren't adding the correct binaries. Still haven't figured out how to install the correct ones yet though 😵

**Comment by benmccann** [comment]
I'm also getting this on my local Linux machine while trying to upgrade `vite-imagetools` to the latest `sharp`. It installs fine, but then fails at runtime when I run `pnpm test`:
```
Error: Could not load the "sharp" module using the linux-x64 runtime
Possible solutions:
- Add platform-specific dependencies:
    npm install --os=linux --cpu=x64 sharp
  or
    npm install --force @img/sharp-linux-x64
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
```

```
  System:
    OS: Linux 6.5 Ubuntu 23.10 23.10 (Mantic Minotaur)
    CPU: (8) x64 Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
  Binaries:
    pnpm: 8.11.0
    Node: 18.16.0
```

**Comment by sawaYch** [comment]
Facing same issue on Vercel. 
Confirmed that my Vercel and local env are running node version = `18.17.1`

I have already install `sharp` by `npm install --force @img/sharp-linux-x64`
When I run locally on a Linux machine (ArchLinux x64) it works fine both on build and runtime; 
while on Vercel build success but the serverless function call throw the following exception at runtime:
```shell
⨯ Error: Could not load the "sharp" module using the linux-x64 runtime
Possible solutions:
- Add platform-specific dependencies:
    npm install --os=linux --cpu=x64 sharp
  or
    npm install --force @img/sharp-linux-x64
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
    at Object.<anonymous> (/var/task/node_modules/sharp/lib/sharp.js:85:9)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Module._load (node:internal/modules/cjs/loader:960:12)
    at r.<computed>.e._load (/var/task/___vc/__launcher/__launcher.js:14:2299)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at w.require (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:11:25815)
    at require (node:internal/modules/cjs/helpers:119:18)
    at Object.<anonymous> (/var/task/node_modules/sharp/lib/constructor.js:10:1) {
  page: '/api/og'
}
```

**Comment by lovell** [comment]
If everything works locally and you're seeing this problem only on Vercel, *please contact Vercel support*.

If you're seeing this problem locally, please [open a new installation issue](https://github.com/lovell/sharp/issues/new?assignees=&labels=installation&projects=&template=installation.md) and answer all the questions. This will help me to gather lots of information from as many people as possible before starting to look for patterns.

**Comment by dresende** [comment]
I'm having this and I'm not using Vercel. Can it be because of an `npm-shrinkwrap.json` created on a MacOS and later on, on Github, the image of my app installs dependencies on a linux?

**Comment by mackermans** [comment]
FYI it's not just on Vercel, we encountered this issue yesterday (caused an incident in prod) on AWS Lambda using the Node 18 runtime because we install latest (0.33.0 in this case) on every deployment. We fixed it by pinning sharp to 0.32.0, but didn't investigate further.

**Comment by lovell** [comment]
I've renamed this issue so it can focus on those experiencing problems on Vercel only. If you are experiencing it anywhere else, please open a new issue.

I genuinely want to help people, and the best way to do that right now is for you to give me lots of information, but not all in the same issue all at once. Lots of separate issues that I can use to find out specifics about each failing environment will allow me to help you to help others. Thank you.

**Comment by JClackett** [comment]
Vercel uses AWS lambda right? so its probably not to do with Vercel itself but AWS lambda in this case

**Comment by acalcutt** [comment]
I am also seeing this error in my tileserver-gl workflow, but I have not had a chance to troubleshoot locally yet.
https://github.com/maptiler/tileserver-gl/pull/1083 .  Looks like it installs sharp but fails with the error mentioned when it tries to run the tests.

This workflow that fails is Ubuntu 20.04/Node18
https://github.com/maptiler/tileserver-gl/actions/runs/7039464819/job/19158648527?pr=1083

One thing I still have to test is the order of sharp/canvas/maplibre-native imports. sometimes the order they are imported seems to make a difference that causes weird errors like this.

**Comment by lovell** [comment]
@acalcutt It looks like you've configured npm to not install optional dependencies by using `--omit=optional`, so please try removing that.

**Comment by devunt** [comment]
Looks like `@vercel/nft` does not pick up corresponding dependency files. We are using it for deployment tree-shaking while not using Vercel itself. We are experiencing the same problem.

Also related: https://github.com/getsentry/sentry-cli/issues/1849

**Comment by lovell** [comment]
Please use either https://vercel.com/help#issues or https://github.com/orgs/vercel/discussions to report issues directly to Vercel.

**Comment by daniel-trevino** [comment]
As a quick fix for now you could install an older version of sharp manually in your app. I tried with `sharp@0.32.6` and it should remove the error from Vercel.

**Comment by devunt** [comment]
> Please use either https://vercel.com/help#issues or https://github.com/orgs/vercel/discussions to report issues directly to Vercel.

Appreciate for the pointing the right direction. However I just wanna let maintainers and others to know that I suspect this is a Vercel (and its library) issue, not a sharp specific issue.

**Comment by devunt** [comment]
After some investigation, if I change https://github.com/lovell/sharp/blob/main/lib/sharp.js#L24 to `require('@img/sharp-darwin-arm64/sharp.node')` `@vercel/nft` collects the corresponding `sharp-darwin-arm64.node` file correctly. I opened the issue at https://github.com/vercel/nft/issues/371.

**Comment by katywings** [comment]
I just updated a project locally from 0.32.6 to 0.33 (using pnpm). v0.33 broke my project with the same error message **Error: Could not load the "sharp" module using the linux-x64 runtime**. I manually had to run `pnpm install -wD @img/sharp-linux-x64`. 

To me it doesn't look like as **At install time, package managers will automatically select prebuilt binaries for the current OS platform and CPU architecture, where available.** is the case anymore and my laptop is definitely not a Vercel server 😅

**Comment by lovell** [comment]
@katywings This issue relates only to Vercel. If you're seeing this problem locally, please [open a new installation issue](https://github.com/lovell/sharp/issues/new?assignees=&labels=installation&projects=&template=installation.md) and answer all the questions. This will help me to gather lots of information from as many people as possible before starting to look for patterns.

**Comment by katywings** [comment]
@lovell I figured it out, I just was pretty unlucky 😅: the node version I was running was v19.8.1 which exactly is missing in the range at https://github.com/lovell/sharp/blob/545e09fad2e10eb9d21c03fe4ad73196e36ffc98/npm/linux-x64/package.json#L32
After switching to node v20.3.0, `pnpm install sharp` automatically installed `@img/sharp-linux-x64` and the error went away.

**Comment by acalcutt** [comment]
I can confirm getting rid of `--omit=optional` fixed my issue, thanks

**Comment by lovell** [comment]
A possible workaround for those deploying Next.js on Vercel and suffering from the problem outlined in https://github.com/vercel/nft/issues/371 is to add something like the following `outputFileTracingIncludes` configuration to your `next.config.js` file (but ensure the paths match those of your app):
```js
module.exports = {
  experimental: {
    outputFileTracingIncludes: {
      '/your/api/route': ['./node_modules/@img/**'],
    },
  },
}
```

https://nextjs.org/docs/pages/api-reference/next-config-js/output#caveats

**Comment by lovell** [comment]
@leppaott This issue relates only to Vercel. If you're seeing this problem locally (or elsewhere), please [open a new installation issue](https://github.com/lovell/sharp/issues/new?assignees=&labels=installation&projects=&template=installation.md) and answer all the questions. This will help me to gather lots of information from as many people as possible before starting to look for patterns.

**Comment by kbourro** [comment]
Same issue after latest version with electron builder

**Comment by lovell** [comment]
I'm going to temporarily lock this issue as it relates only and specifically to a problem with Vercel that is being tracked at https://github.com/vercel/nft/issues/371 - please subscribe to that issue and/or contact Vercel support for updates.

For anyone else coming here, please [open a new installation issue](https://github.com/lovell/sharp/issues/new?assignees=&labels=installation&projects=&template=installation.md) and answer all the questions. I need your answers to these questions as a minimum starting point to be able to try to help you.

**Comment by lovell** [comment]
The https://github.com/vercel/nft/issues/371 issue has now been fixed. If you are deploying to Vercel (or using Next.js standalone mode) please ensure you upgrade your dependencies to use `@vercel/nft` version 0.26.1 or later.

For anyone else coming here experiencing similar symptoms, please [open a new installation issue](https://github.com/lovell/sharp/issues/new?assignees=&labels=installation&projects=&template=installation.md) and answer all the questions.

**Comment by hewar-saber** [comment]
> sharp@0.32.6

This worked for me, I had the same issue with Next.js on Vercel and Netlify.

**Comment by hewar-saber** [comment]
> As a quick fix for now you could install an older version of sharp manually in your app. I tried with `sharp@0.32.6` and it should remove the error from Vercel.


sharp@0.32.6

This worked for me, I had the same issue with Next.js on Vercel and Netlify.

**Comment by JClackett** [comment]
I wasn't using @vercel/nft, and its still error in vercel for me using latest versions

**Comment by lovell** [comment]
@JClackett I think Vercel uses `@vercel/nft` under the hood so perhaps they have not yet updated their build infrastructure with the latest version. You'll need to ask Vercel support about this.

**Comment by slavanossar** [comment]
Was also experiencing this issue and updating Node fixed the issue (`18.16.1` > `18.19.01).

**Comment by aoioioi** [comment]
Not working for me deployed on render. Using node 18.19.0. Rolling back to 0.32.6 fixed the issue for me

**Comment by dalyr95** [comment]
Same issue on AWS, not vercel. Node 18.17.0 - x64, 0.32.6 rollback works

**Comment by lovell** [comment]
There was a problem affecting both Next.js standalone mode and Vercel deployments, which has now been fixed - please ensure `@vercel/nft` version 0.26.1 or later is being used - see https://github.com/vercel/nft/issues/371 for details.

Lots of people are running into https://github.com/npm/cli/issues/4828 so please check that one first if you're sharing lockfiles between multiple platforms.

If you're still having problems, please [open a new installation issue](https://github.com/lovell/sharp/issues/new?assignees=&labels=installation&projects=&template=installation.md) and answer all the questions. I need your answers to these questions as a minimum starting point to be able to try to help you.

I'm going to lock this issue again.
